<b>¡Hola Oscar!</b>

Mi nombre es Alejandro Abia y tengo el gusto de revisar tu proyecto.

A continuación, encontrarás mis comentarios en celdas pintadas de tres colores (verde, amarillo y rojo), a manera de semáforo. Por favor, <b>no las borres ni muevas de posición</b> mientras dure el proceso de revisión.

<div class="alert alert-block alert-success">
<b>Éxito</b> <a class="tocSkip"></a>
En celdas verdes encontrarás comentarios en relación a tus aciertos y fortalezas.
</div>
<div class="alert alert-block alert-warning">
<b>Atención</b> <a class="tocSkip"></a>
Utilizaré el color amarillo para llamar tu atención, expresar algo importante o compartirte alguna idea de valor.
</div>
<div class="alert alert-block alert-danger">
<b>A resolver</b> <a class="tocSkip"></a>
En rojo emitiré aquellos puntos que podrían impedir que el proyecto se ejecute correctamente. No son errores, sino oportunidades importantes de mejora.
</div>
<div class="alert alert-block alert-info">
<b>Comentario estudiante</b> <a class="tocSkip"></a>
Si durante la revisión deseas dejarme algún comentario, por favor utiliza celdas azules como esta.
</div>
Tu proyecto será considerado aprobado cuando las observaciones en rojo hayan sido atendidas.  
¡Empecemos!

<div style="background-color: #F4C49C; padding: 15px; border-radius: 12px;">

# 📘 Proyecto Producto Literario  

## 🧩 Contexto Introductorio  

Durante la pandemia de <b>COVID-19</b>, los hábitos de consumo de contenido digital cambiaron drásticamente. Las personas comenzaron a pasar más tiempo en casa, dedicando parte de su tiempo libre a la lectura como una forma de entretenimiento y desarrollo personal.<br>

Este fenómeno impulsó el crecimiento de múltiples <b>startups literarias</b> y aplicaciones orientadas a mejorar la experiencia de los lectores, ofreciendo recomendaciones personalizadas, reseñas y calificaciones de libros.<br>

Como parte de este contexto, una empresa emergente del sector tecnológico-literario ha proporcionado una base de datos con información sobre:<br>
- Libros publicados y sus características generales.<br>
- Autores y editoriales asociadas.<br>
- Calificaciones y reseñas otorgadas por los usuarios.<br>

El propósito general de este proyecto es utilizar dicha información para <b>comprender las dinámicas del mercado editorial digital</b> y <b>apoyar el desarrollo de un nuevo producto de lectura</b> que se base en las preferencias y comportamientos reales de los usuarios.

</div>


<div style="background-color: #d8c7f5; padding: 10px; font-weight: bold;">

## PASO 1: 🎯 OBJETIVOS DEL ESTUDIO

</div>


<div style="background-color: #FBE8D3; padding: 15px; border-radius: 10px;">

El presente análisis tiene como propósito <b>examinar la base de datos del servicio literario</b> con el fin de identificar tendencias, patrones y relaciones significativas entre los libros, sus autores, las editoriales y la interacción de los usuarios.<br>

En particular, se busca:<br>
🔹 Analizar la <b>distribución y características</b> de los libros (páginas, fechas de publicación, editoriales).<br>
🔹 Evaluar la <b>participación y comportamiento de los usuarios</b> a través de calificaciones y reseñas.<br>
🔹 Identificar <b>autores y editoriales con mejor desempeño</b> en función de sus métricas.<br>
🔹 Generar <b>hallazgos accionables</b> que contribuyan a la toma de decisiones sobre el diseño del nuevo producto.<br>

En conjunto, este estudio permitirá <b>obtener una visión integral del ecosistema literario digital</b> y fundamentar estrategias orientadas al crecimiento, basadas en datos y evidencia empírica.

</div>


<div style="background-color: #d8c7f5; padding: 10px; font-weight: bold;">

## 🎯 PASO 2: CONEXION A LA BASE DE DATOS

</div>

<div style="background-color:#FBE8D3; padding:15px; border-radius:12px; text-align:justify;">
El siguiente paso consiste en establecer la conexión con la base de datos que contiene la información sobre 
<b>libros, autores, editoriales, calificaciones y reseñas</b>. 
<br><br>
Para ello, se empleará el motor de conexión <b>SQLAlchemy</b> junto con <b>pandas</b>, lo que permitirá ejecutar consultas SQL directamente desde Python y transformar los resultados en <b>DataFrames</b> para su análisis posterior.
<br><br>
Esta integración facilita combinar la potencia de las consultas SQL con la flexibilidad de análisis y visualización que ofrece Python, optimizando el flujo de trabajo en proyectos de análisis de datos.
</div>

In [1]:
# importar librerías
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'practicum_student',         # nombre de usuario
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # contraseña
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # puerto de conexión
             'db': 'data-analyst-final-project-db'}          # nombre de la base de datos

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})



<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
En la celda [1], configuraste la conexión con SQLAlchemy usando <code>create_engine</code> y forzando <code>sslmode='require'</code>. Es un buen detalle porque asegura el canal cifrado hacia PostgreSQL y estandariza el acceso desde pandas. Esto hace tu flujo reproducible y claro. Vale la pena mantener este patrón en futuros proyectos para que la extracción de datos sea consistente y segura.
</div>


<div style="background-color: #d8c7f5; padding: 10px; font-weight: bold;">

## PASO 3: 📦 REVISION DE LAS TABLAS

</div>


<div style="background-color:#FBE8D3; padding:15px; border-radius:12px; text-align:justify;">
Antes de realizar consultas específicas, es fundamental <b>explorar la estructura y contenido</b> de las tablas disponibles en la base de datos. 
<br><br>
Esto permitirá comprender la naturaleza de los datos, sus relaciones y los campos clave que se utilizarán en las consultas SQL del análisis principal.
</div>


In [2]:
# Visualización de la tabla books
query = """
SELECT *
FROM books
LIMIT 5;
"""
books = pd.read_sql(query, con=engine)
display(books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
Las celdas [2] a [6] muestran una exploración inicial con <code>SELECT * ... LIMIT 5</code>. Es una práctica útil para orientarte con los campos: en <code>books</code> confirmaste <code>num_pages</code>, <code>publication_date</code> y <code>publisher_id</code>; en <code>ratings</code> y <code>reviews</code> validaste la presencia de <code>username</code> para relacionar actividad de usuarios. Este vistazo rápido facilita diseñar consultas más precisas después.
</div>

In [3]:
# Visualización de la tabla authors
query = """
SELECT *
FROM authors
LIMIT 5;
"""
authors = pd.read_sql(query, con=engine)
display(authors)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [4]:
# Visualización de la tabla publishers
query = """
SELECT *
FROM publishers
LIMIT 5;
"""
publishers = pd.read_sql(query, con=engine)
display(publishers)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [5]:
# Visualización de la tabla ratings
query = """
SELECT *
FROM ratings
LIMIT 5;
"""
ratings = pd.read_sql(query, con=engine)
display(ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [6]:
# Visualización de la tabla reviews
query = """
SELECT *
FROM reviews
LIMIT 5;
"""
reviews = pd.read_sql(query, con=engine)
display(reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<div style="background-color: #d8c7f5; padding: 10px; font-weight: bold;">

## PASO 4: 🧩 CONSULTA DE TAREAS CON SQL

</div>

<div style="background-color: #F4ECD8; padding: 15px; border-radius: 13px;">
    
### **🔹 4.1. ENCUENTRA EL NUMERO DE LIBROS PUBLICADOS DESPUES DEL 1 DE ENERO DE 2000**    

</div>

In [7]:
query = """
SELECT 
    COUNT(book_id) AS published_books
FROM 
    books
WHERE
    publication_date > '2000-01-01';
"""
published_books = pd.read_sql(query, con=engine)
display(published_books)

published_books
0              819

<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
La consulta de la celda [7] para contar libros publicados después de <code>'2000-01-01'</code> está bien estructurada, con alias claro (<code>published_books</code>) y filtro explícito. Esta precisión facilita reutilizar el resultado y deja rastro claro de la lógica temporal aplicada.
</div>

In [8]:
query = """
SELECT
    MIN(publication_date) AS first_publication,
    MAX(publication_date) AS last_publication
FROM
    books;
"""

publication_range = pd.read_sql(query, con=engine)
display(publication_range)


first_publication last_publication
0        1952-12-01       2020-03-31

<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
La celda [8] que obtiene <code>MIN</code> y <code>MAX</code> de <code>publication_date</code> es un buen contexto temporal: anclar el rango (1952-12-01 a 2020-03-31) ayuda a interpretar cualquier métrica de ritmo editorial posterior. Este paso previo ordena la narrativa del análisis.
</div>

In [9]:
query = """
SELECT
    COUNT(book_id) AS total_books_post_2000,
    ROUND(COUNT(book_id) / (2020 - 2000), 2) AS avg_books_per_year
FROM
    books
WHERE
    publication_date > '2000-01-01';
"""

avg_books_post_2000 = pd.read_sql(query, con=engine)
display(avg_books_post_2000)


total_books_post_2000  avg_books_per_year
0                    819                40.0

<div style="background-color: #EAD6D1; padding: 15px; border-radius: 10px;">
    
#### **CONCLUSIONES 4.1: ENCUENTRA EL NUMERO DE LIBROS PUBLICADOS DESPUES DEL 1 DE ENERO DE 2000**

Entre los años <b>2000 y 2020</b>, se registraron <b>819 publicaciones</b>, lo que equivale a un promedio de <b>40 libros por año</b>. Este resultado refleja un <b>ritmo editorial sostenido</b> a lo largo de las dos primeras décadas del siglo XXI.<br><br>
Sin embargo, determinar si este volumen es <b>alto o bajo</b> requeriría compararlo con <b>promedios del sector editorial</b> y con la <b>región geográfica</b> a la que pertenece el catálogo analizado. Aun así, los datos sugieren una tendencia estable que podría estar relacionada con la <b>transición hacia modelos de publicación digital</b> y la diversificación de autores y editoriales en este periodo.

</div>

<div style="background-color: #F4ECD8; padding: 15px; border-radius: 13px;">
    
### **🔹 4.2. ENCUENTRA EL NUMERO DE RESEÑAS DE USUARIOS Y LA CALIFICACION PROMEDIO DE CADA LIBRO**    

</div>

In [10]:
query = """
SELECT
    ratings.book_id AS Book_ID,
    COUNT(reviews.review_id) AS Total_Reviews,
    ROUND(AVG(ratings.rating), 2) AS Average_Book_Rating
FROM
    ratings
    LEFT JOIN reviews ON ratings.book_id = reviews.book_id
GROUP BY
    ratings.book_id
ORDER BY
    Average_Book_Rating DESC;
"""
ratings_and_reviews = pd.read_sql(query, con=engine)
display(ratings_and_reviews)

book_id  total_reviews  average_book_rating
0        518              4                 5.00
1        732              4                 5.00
2        347              6                 5.00
3        610              4                 5.00
4        330              2                 5.00
..       ...            ...                  ...
995      915             12                 2.25
996      371              4                 2.00
997      316              4                 2.00
998      202              9                 2.00
999      303              4                 1.50

[1000 rows x 3 columns]

<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
Las conclusiones que acompañan a [10] sintetizan bien el objetivo y el criterio del <code>LEFT JOIN</code> (incluir libros con calificación aunque no tengan reseñas). Esta claridad metodológica ayuda a quien lee a entender por qué uniones de este tipo son útiles para no perder registros relevantes en análisis descriptivos.
</div>

<div style="background-color: #EAD6D1; padding: 15px; border-radius: 10px;">
    
#### **CONCLUSIONES 4.2: ENCUENTRA EL NUMERO DE RESEÑAS DE USUARIOS Y LA CALIFICACION PROMEDIO DE CADA LIBRO**

A partir del análisis de las tablas <i>ratings</i> y <i>reviews</i>, se obtuvo el número total de reseñas registradas y la calificación promedio correspondiente a cada libro. 
Para asegurar que todos los libros con calificación fueran incluidos, se utilizó un <code>LEFT JOIN</code>, lo que permite incorporar aquellos títulos que poseen valoraciones pero no necesariamente comentarios asociados. <br><br>
En total se identificaron <b>1,000 libros</b> con calificaciones registradas. Los resultados muestran que existen múltiples libros con una calificación promedio máxima de <b>5.00</b>, mientras que los valores más bajos rondan entre <b>1.50</b> y <b>2.00</b>, lo que indica una distribución amplia en la percepción de los lectores. <br><br>
El uso de la función agregada <code>AVG()</code> permitió calcular la media de las valoraciones por cada identificador de libro (<i>book_id</i>), y con <code>COUNT()</code> se determinó el número total de reseñas asociadas. 
Este enfoque ofrece una visión clara del comportamiento general de los usuarios en relación con los títulos disponibles y facilita identificar tanto los libros mejor valorados como aquellos con baja participación o calificaciones desfavorables. 

</div>

<div style="background-color: #F4ECD8; padding: 15px; border-radius: 13px;">
    
### **🔹 4.3. IDENTIFICA LA EDITORIAL QUE HA PUBLICADO EL MAYOR NUMERO DE LIBROS CON MAS DE 50 PAGINAS**    

</div>

In [11]:
query = """
SELECT
    books.publisher_id AS Publisher_Id,
    publishers.publisher AS Publisher_Name,
    COUNT(books.book_id) AS Books_Over_50_Pages
FROM
    books
    LEFT JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE
    books.num_pages > 50
GROUP BY
    books.publisher_id,
    publishers.publisher
ORDER BY
    Books_Over_50_Pages DESC;
"""
top_publishers = pd.read_sql(query, con=engine)
display(top_publishers)

publisher_id                         publisher_name  books_over_50_pages
0             212                          Penguin Books                   42
1             309                                Vintage                   31
2             116               Grand Central Publishing                   25
3             217                       Penguin Classics                   24
4              33                       Ballantine Books                   19
..            ...                                    ...                  ...
329           148          Harvard Business Review Press                    1
330           302               Tyndale House Publishers                    1
331            66                        Corgi Childrens                    1
332           230  Pocket Books / Simon & Schuster  Inc.                    1
333           199                          Orchard Books                    1

[334 rows x 3 columns]

<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
La celda [11] relaciona <code>books</code> con <code>publishers</code> y filtra <code>num_pages &gt; 50</code> antes de agrupar. Es una secuencia clara: filtro → join → agregación. Además, los alias (<code>Books_Over_50_Pages</code>) hacen el resultado autocontenible y fácil de leer.
</div>

<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
En la conclusión de [11], interpretaste el ranking de editoriales con más títulos &gt; 50 páginas y conectaste el hallazgo con su posible posicionamiento y catálogo. Esa conexión entre resultado y uso práctico es valiosa para producto: aterriza el dato en una decisión.
</div>

<div style="background-color: #EAD6D1; padding: 15px; border-radius: 10px;">
    
#### **CONCLUSIONES 4.3: IDENTIFICA LA EDITORIAL QUE HA PUBLICADO EL MAYOR NUMERO DE LIBROS CON MAS DE 50 PAGINAS**

Del análisis de las tablas <i>books</i> y <i>publishers</i>, se identificó qué editoriales han publicado el mayor número de libros con más de 50 páginas. 
La consulta empleó una combinación mediante <code>LEFT JOIN</code> para relacionar los identificadores de editorial (<i>publisher_id</i>) 
con su nombre correspondiente y un filtro en la cláusula <code>WHERE</code> que limita los resultados a obras con más de 50 páginas. <br><br>
Los resultados muestran que <b>Penguin Books</b> encabeza la lista con <b>42 títulos</b>, seguida de <b>Vintage</b> (31) y <b>Grand Central Publishing</b> (25). 
Este patrón sugiere una clara concentración de publicaciones extensas en editoriales de gran reconocimiento, 
posiblemente vinculadas a catálogos de literatura clásica o de alta calidad editorial. <br><br>
Este tipo de análisis permite identificar a las editoriales con mayor producción en títulos de formato largo, 
lo cual puede ser útil para evaluar la estrategia de publicación y el posicionamiento en segmentos de contenido más sustantivo.

</div>

<div style="background-color: #F4ECD8; padding: 15px; border-radius: 13px;">
    
### **🔹 4.4. IDENTIFICA AL AUTOR QUE TIENE LA MAS ALTA CALIFICACION PROMEDIO DEL LIBRO**    

</div>

In [12]:
query = """
SELECT
    authors.author_id AS Author_id,
    authors.author AS Author_Name,
    ROUND(AVG(ratings.rating),2) AS Rating_Book_Average
FROM
    ratings
    LEFT JOIN books ON ratings.book_id = books.book_id
    LEFT JOIN authors ON books.author_id = authors.author_id
GROUP BY
    authors.author_id,
    authors.author
HAVING
    COUNT(ratings.rating) >= 50
ORDER BY
    Rating_Book_Average DESC
LIMIT 10;
"""
top_author_by_rating_book = pd.read_sql(query, con=engine)
top_author_by_rating_book

author_id                        author_name  rating_book_average
0        130                     Diana Gabaldon                 4.30
1        236         J.K. Rowling/Mary GrandPré                 4.29
2          3                    Agatha Christie                 4.28
3        402  Markus Zusak/Cao Xuân Việt Khương                 4.26
4        240                     J.R.R. Tolkien                 4.24
5        499           Roald Dahl/Quentin Blake                 4.21
6        376                  Louisa May Alcott                 4.20
7        498                       Rick Riordan                 4.13
8         39                      Arthur Golden                 4.11
9        542                       Stephen King                 4.01

<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
En la celda [12], el uso de <code>HAVING COUNT(ratings.rating) &gt;= 50</code> para asegurar un mínimo de evaluaciones por autor es un gran acierto. Esta decisión reduce el sesgo por tamaños de muestra pequeños y fortalece la validez de la comparación entre autores. Además, el redondeo y el <code>LIMIT 10</code> cierran bien el reporte.
</div>

<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
La conclusión de [12] destaca el uso de <code>HAVING</code> para asegurar validez estadística y lo explica en lenguaje claro. Ese énfasis en supuestos y tamaños mínimos de muestra es una fortaleza metodológica que eleva la calidad del análisis.
</div>


<div style="background-color: #EAD6D1; padding: 15px; border-radius: 10px;">
    
#### **CONCLUSIONES 4.4: IDENTIFICA AL AUTOR QUE TIENE LA MAS ALTA CALIFICACION PROMEDIO DEL LIBRO**

El análisis de las tablas <i>ratings</i>, <i>books</i> y <i>authors</i> permitió identificar a los autores con la calificación promedio más alta, 
considerando únicamente aquellos cuyos libros han recibido <b>al menos 50 calificaciones</b>. 
Para ello se aplicó una combinación de las funciones agregadas <code>AVG()</code> y <code>COUNT()</code> junto con la cláusula <code>HAVING</code>, 
que filtra los grupos resultantes del <code>GROUP BY</code> para incluir solo los autores con un número representativo de evaluaciones. <br><br>
Los resultados muestran a <b>Diana Gabaldon</b> como la autora con la mayor calificación promedio (<b>4.30</b>), 
seguida por <b>J.K. Rowling y Mary GrandPré</b> (4.29), <b>Agatha Christie</b> (4.28) y <b>J.R.R. Tolkien</b> (4.24). 
Estos valores reflejan una fuerte aprobación de los lectores y una consistencia positiva en la calidad percibida de sus obras. <br><br>
La aplicación de <code>HAVING COUNT(ratings.rating) >= 50</code> fue clave para garantizar la validez estadística de los resultados, 
evitando sesgos provocados por autores con escasas calificaciones. 
Este enfoque demuestra la importancia del filtrado posterior al agrupamiento en SQL para obtener conclusiones sólidas y comparativas.

</div>


<div style="background-color: #F4ECD8; padding: 15px; border-radius: 13px;">
    
### **🔹 4.5. NUMERO PROMEDIO DE RESEÑAS ENTRE LOS USUARIOS QUE CALIFICARON MAS DE 50 LIBROS**    

</div>

In [13]:
# Obtener la lista de usuarios que calificaron más de 50 libros.
query = """
SELECT
    ratings.username AS Username,
    COUNT(reviews.review_id) AS Total_Reviews
FROM
    ratings
LEFT JOIN reviews ON ratings.username = reviews.username
GROUP BY
    ratings.username
HAVING
    COUNT(ratings.rating) > 50
ORDER BY
    Total_Reviews DESC;
"""

top_review_users = pd.read_sql(query, con=engine)
display(top_review_users)

username  total_reviews
0       sfitzgerald           1540
1        martinadam           1512
2         richard89           1430
3           susan85           1421
4    jennifermiller           1325
..              ...            ...
155        dwayne09            370
156        daniel19            341
157        joseph14            330
158        robert00            261
159  millernicholas            160

[160 rows x 2 columns]

In [14]:
## Realizar la subconsulta: Promedio de reseñas entre los usuarios que calificaron más de 50 libros
query_avg = """
SELECT
    ROUND(AVG(user_reviews.Total_Reviews), 2) AS Avg_Reviews_Among_Top_Users
FROM (
    SELECT
        ratings.username,
        COUNT(reviews.review_id) AS Total_Reviews
    FROM
        ratings
    LEFT JOIN reviews ON ratings.username = reviews.username
    GROUP BY
        ratings.username
    HAVING
        COUNT(ratings.rating) > 50
) AS user_reviews;
"""

avg_reviews = pd.read_sql(query_avg, con=engine)
display(avg_reviews)



avg_reviews_among_top_users
0                       719.76

<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
El cierre general tras [14] está bien hilado: conectas contexto, objetivos, consultas y hallazgos con implicaciones de producto (recomendación, participación, fidelización). Esta narrativa hace que el trabajo no se quede en métricas aisladas y muestre su valor aplicado.
</div>

<div style="background-color: #EAD6D1; padding: 15px; border-radius: 10px;">
    
#### **CONCLUSIONES 4.5: NUMERO PROMEDIO DE RESEÑAS ENTRE LOS USUARIOS QUE CALIFICARON MAS DE 50 LIBROS**

El análisis permitió calcular el número promedio de reseñas de texto entre los usuarios más activos, 
es decir, aquellos que han calificado más de <b>50 libros</b>. 
Para realizarlo, se aplicó un enfoque basado en la combinación de las tablas <i>ratings</i> y <i>reviews</i> mediante un <code>LEFT JOIN</code>, 
vinculando las calificaciones con las reseñas escritas por cada usuario. Posteriormente, los datos fueron agrupados 
por nombre de usuario con <code>GROUP BY</code> y filtrados con <code>HAVING COUNT(ratings.rating) &gt; 50</code>, 
lo que permitió concentrarse en los usuarios con un nivel de participación elevado.<br><br>
El uso de la cláusula <code>HAVING</code> fue fundamental, ya que permitió aplicar la condición de filtro sobre los resultados agregados 
posteriores al agrupamiento, algo que no habría sido posible con <code>WHERE</code>. 
Con esta estructura, se obtuvo el número total de reseñas de texto por cada usuario activo, y posteriormente, 
se calculó el promedio global mediante la función <code>AVG()</code> en una subconsulta.<br><br>
El resultado final mostró que los usuarios que calificaron más de 50 libros generaron un promedio de aproximadamente 
<b>719.76 reseñas de texto</b> por usuario. 
Este hallazgo indica un fuerte nivel de compromiso entre los usuarios más activos, quienes no solo califican una gran cantidad de libros, 
sino que también aportan contenido cualitativo adicional a través de reseñas detalladas. 
Esto refuerza la idea de que existe una correlación positiva entre la frecuencia de participación y la contribución de valor a la comunidad de lectores.

</div>

<div style="background-color: #d8c7f5; padding: 10px; font-weight: bold;">

## PASO 5: 🧭 CONCLUSION GENERAL

</div>

<div style="background-color:#F4C49C; padding:12px; border-radius:10px;">

<p>El análisis realizado permitió examinar de manera integral el comportamiento del mercado editorial digital, integrando datos sobre libros, autores, editoriales y usuarios activos para revelar patrones y relaciones clave. Mediante el uso de consultas <b>SQL</b>, se identificaron tendencias que explican las dinámicas del ecosistema literario contemporáneo y su relación con las preferencias de lectura en entornos digitales.</p>

<p>Desde el <b>contexto introductorio</b>, se estableció que la digitalización impulsada por la pandemia de <b>COVID-19</b> favoreció el crecimiento de múltiples empresas emergentes del sector literario y de aplicaciones orientadas a mejorar la experiencia de los lectores. En este marco, el proyecto logró cumplir los objetivos planteados al:</p>

<ul>
  <li>Analizar la <b>distribución temporal y estructural</b> de las publicaciones, evidenciando un ritmo sostenido de producción editorial entre 2000 y 2020.</li>
  <li>Considerar la posible discrepancia entre calificaciones y reseñas, empleando un <b>LEFT JOIN</b> para garantizar que todos los libros valorados fueran incluidos, aun cuando no tuvieran comentarios registrados.</li>
  <li>Identificar las <b>editoriales con mayor producción</b> de obras extensas, destacando a <i>Penguin Books</i> y <i>Vintage</i> por su liderazgo en títulos de más de 50 páginas.</li>
  <li>Determinar los <b>autores con mejor calificación promedio</b>, encabezados por <b>Diana Gabaldon</b> y <b>J.K. Rowling</b>, con promedios superiores a 4.2, lo que refleja una alta aceptación entre los lectores.</li>
  <li>Evaluar el <b>nivel de participación de los usuarios</b>, concluyendo que quienes calificaron más de 50 libros registraron un promedio de <b>719.76 reseñas de texto</b>, demostrando un compromiso significativamente elevado.</li>
</ul>

<p>En conjunto, los resultados revelan un ecosistema literario digital caracterizado por la interacción entre cantidad y calidad: los usuarios más activos no solo califican con frecuencia, sino que también aportan reseñas detalladas, generando un valor agregado para la comunidad lectora. Del mismo modo, la concentración de publicaciones en editoriales reconocidas y la preferencia por autores consolidados evidencian un mercado literario que <b>privilegia la calidad de las obras y el cuidado de la experiencia del lector</b>.</p>

<p>Este estudio evidenció cómo el análisis de <b>información editorial</b> puede convertirse en conocimiento estratégico para guiar decisiones, optimizar catálogos y crear productos alineados con los hábitos reales de lectura. En otras palabras, el uso de <b>SQL</b> no solo permitió obtener métricas descriptivas, sino también generar conocimiento aplicable al desarrollo de nuevas oportunidades dentro del mercado editorial digital.</p>

<p>Finalmente, el análisis de la participación de los usuarios refuerza la importancia de comprender los <b>patrones de interacción y fidelización</b> en el diseño de un nuevo producto literario digital. Estos hallazgos pueden guiar a las empresas del sector a implementar estrategias de recomendación personalizadas, fomentar la participación activa de la comunidad y consolidar una <b>oferta basada en datos</b> que conecte de manera más efectiva con los intereses y comportamientos reales de los lectores.</p>

</div>

<div class="alert alert-block alert-success">
<b>Comentario final</b> <a class="tocSkip"></a><br>
¡Muy buen trabajo, Oscar! A lo largo del proyecto mostraste fortalezas muy claras:<br><br>
• Contextualizaste muy bien el problema y definiste objetivos concretos y accionables.<br>
• Configuraste la conexión con SQLAlchemy usando SSL, un buen estándar de seguridad y portabilidad.<br>
• Realizaste una exploración inicial ordenada de las tablas, confirmando campos clave para el análisis.<br>
• Usaste alias claros y funciones agregadas (<code>COUNT</code>, <code>MIN</code>, <code>MAX</code>, <code>AVG</code>) con precisión.<br>
• Incluiste el rango temporal de publicaciones, lo que dio marco a los resultados posteriores.<br>
• Aplicaste <code>LEFT JOIN</code> para no perder libros con calificación aunque no tuvieran reseñas, justificándolo en el texto.<br>
• Construiste rankings útiles (editoriales y autores) con filtros razonables (páginas &gt; 50, mínimo de 50 calificaciones).<br>
• Usaste <code>HAVING</code> de forma correcta para asegurar tamaños de muestra adecuados en el análisis por autor.<br>
• Ordenaste y redondeaste resultados para que fueran legibles y directos para decisión.<br>
• Acompañaste las consultas con conclusiones claras y conectadas al diseño de producto.<br>
• Mantuvista coherencia entre código y narrativa, facilitando la lectura global del notebook.<br>
• Estructuraste el cuaderno por pasos (objetivos, conexión, revisión, consultas, conclusiones), lo que hizo muy fácil seguir el hilo.<br>
• Escogiste métricas pertinentes (conteos, promedios, top-N) para responder preguntas concretas.<br>
• Mostraste atención a la representatividad (umbral de 50 ratings), un criterio clave en análisis descriptivo.<br>
• Cerraste con una síntesis que traduce los hallazgos en insumos útiles para roadmap de producto.<br><br>
¡Felicidades!
</div>